File: Solutions/40. Combination Sum II/40. Combination Sum II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res=new ArrayList<>();
        List<Integer> path=new ArrayList<>();
        Arrays.sort(candidates);
        foo(candidates,target,res,path,0);
        return res;
    }
    private void foo(int[] c,int t,List<List<Integer>> res,List<Integer> path,int begin){
        if(t<=0){
            if(t==0){
                res.add(new ArrayList<Integer>(path));
                return;
            }
            return;
        }
        for(int i=begin;i<c.length;++i){
            if(i>begin && c[i]==c[i-1])
                continue;
            path.add(c[i]);
            foo(c,t-c[i],res,path,i+1);
            path.remove(path.size()-1);
        }
    }
}

File: Solutions/40. Combination Sum II/40. Combination Sum II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def foo(c,t,res,path,begin):
            if t<=0:
                if t==0:
                    res.append(path.copy())
                    return
                return
            for i in range(begin,len(c)):
                if i>begin and c[i]==c[i-1]:
                    continue
                path.append(c[i])
                foo(c,t-c[i],res,path,i+1)
                path.pop()
        candidates.sort()
        path,res=[],[]
        foo(candidates,target,res,path,0)
        return res
                    